contracts/cw-token-swap/src/contract.rs

```diff
@@ -0,0 +1,247 @@
+};
+
+const CONTRACT_NAME: &str = "crates.io:cw-token-swap";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let counterparty_one = msg.counterparty_one.into_checked(deps.as_ref())?;
+    let counterparty_two = msg.counterparty_two.into_checked(deps.as_ref())?;
+
+    COUNTERPARTY_ONE.save(deps.storage, &counterparty_one)?;
+    COUNTERPARTY_TWO.save(deps.storage, &counterparty_two)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("counterparty_one", counterparty_one.address)
+        .add_attribute("counterparty_two", counterparty_two.address))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(msg) => execute_receive(deps, info.sender, msg),
+        ExecuteMsg::Fund {} => execute_fund(deps, info),
+        ExecuteMsg::Withdraw {} => execute_withdraw(deps, info),
+    }
+}
+
+struct CounterpartyResponse<'a> {
+    pub counterparty: CheckedCounterparty,
+    pub other_counterparty: CheckedCounterparty,
+    pub storage: Item<'a, CheckedCounterparty>,
+}
+
+fn get_counterparty<'a>(
+    deps: Deps,
+    sender: &Addr,
+) -> Result<CounterpartyResponse<'a>, ContractError> {
+    let counterparty_one = COUNTERPARTY_ONE.load(deps.storage)?;
+    let counterparty_two = COUNTERPARTY_TWO.load(deps.storage)?;
+
+    let (counterparty, other_counterparty, storage) = if *sender == counterparty_one.address {
+        (counterparty_one, counterparty_two, COUNTERPARTY_ONE)
+    } else if *sender == counterparty_two.address {
+        (counterparty_two, counterparty_one, COUNTERPARTY_TWO)
+    } else {
+        // Contract may only be funded by a counterparty.
+        return Err(ContractError::Unauthorized {});
+    };
+
+    Ok(CounterpartyResponse {
+        counterparty,
+        other_counterparty,
+        storage,
+    })
+}
+
+/// Accepts funding from COUNTERPARTY for the escrow. Distributes
+/// escrow funds if both counterparties have funded the contract.
+///
+/// NOTE: The caller must verify that the denom of PAID is correct.
+fn do_fund(
+    deps: DepsMut,
+    counterparty: CheckedCounterparty,
+    paid: Uint128,
+    expected: Uint128,
+    other_counterparty: CheckedCounterparty,
+    storage: Item<CheckedCounterparty>,
+) -> Result<Response, ContractError> {
+    if paid != expected {
+        return Err(ContractError::InvalidAmount {
+            expected,
+            actual: paid,
+        });
+    }
+
+    if counterparty.provided {
```

Comment:
nit: this can be the first check

Reply:
Great point. Likely the more relevant of the two errors. :)